projects/engine-vscode/src/webview.ts

```diff
@@ -0,0 +1,140 @@
+  panel?: WebviewPanel
+  options: WebviewOptions
+
+  constructor(profile: Profile & ExternalProfile, options: WebviewOptions) {
+    super(profile)
+    this.options = options
+  }
+
+  setOptions(options: WebviewOptions) {
+    super.setOptions(options)
+  }
+
+  protected send(message: Partial<Message>): void {
+    if (this.panel) {
+      this.panel.webview.postMessage(message)
+    }
+  }
+
+  protected connect(url: string): void {
+    if (this.options.context) {
+      const { extensionPath } = this.options.context
+      this.panel = createWebview(this.profile, url, extensionPath, this.options.column)
+      this.panel.webview.onDidReceiveMessage(msg => this.getMessage(msg))
+      this.options.context.subscriptions.push(this.panel)
+    } else {
+      throw new Error(`WebviewPlugin "${this.name}" `)
+    }
+  }
+
+  protected disconnect(): void {
+    if (this.panel) {
+      this.panel.dispose()
+    }
+  }
+
+}
+
+
+/** Create a webview */
+export function createWebview(profile: Profile, url: string, extensionPath: string, column: ViewColumn) {
+  const { protocol, path } = parseUrl(url)
+  const { ext } = parsePath(path)
+  const isRemote = protocol === 'https:' || protocol === 'http:'
+  const baseUrl = isRemote
+    ? ext === '.html' ? parsePath(url).dir  : url
+    : ext === '.html' ? parsePath(path).dir : url
+
+  const panel = window.createWebviewPanel(
+    profile.name,
+    profile.displayName || profile.name,
+    column || window.activeTextEditor?.viewColumn || ViewColumn.One,
+    {
+      enableScripts: true,
+      localResourceRoots: isRemote ? [] : [Uri.file(join(extensionPath, baseUrl))]
+    })
+
+
+  isRemote
+    ? setRemoteHtml(panel.webview, baseUrl)
+    : setLocalHtml(panel.webview, join(extensionPath, baseUrl))
+
+  // TODO: If dev mode update on change
+  // if (devMode) {
+  //   watch(index).on('change', updateWebview)
+  // }
+
+  return panel
+}
+
+/** Fetch remote ressource with http */
+function fetch(url: string): Promise<string> {
+  return new Promise((resolve, reject) => {
+    get(url, res => {
+      let text = ''
+      res.on('data', data => text += data)
+      res.on('end', _ => resolve(text))
+      res.on('error', err => reject(err))
+    })
+  })
+}
+
+
+/** Get code from remote source */
+async function setRemoteHtml(webview: Webview, baseUrl: string) {
+  const matchLinks = /(href|src)="([^"]*)"/g
+  const index = `${baseUrl}/index.html`
+
+
+  // Vscode requires URI format from the extension root to work
+  const toRemoteUrl = (_: any, prefix: 'href' | 'src', link: string) => {
+    // For <base href="#" />
+    if (link === '#') {
+      return `${prefix}="${link}"`
+    }
+    // For scripts & links
+    const path = join(baseUrl, link)
+    return `${prefix}="${path}"`
+  }
+
+  const html = await fetch(index)
+  webview.html = html.replace(matchLinks, toRemoteUrl)
+}
+
+
+/** Get code from local source */
+async function setLocalHtml(webview: Webview, baseUrl: string) {
```

Comment:
I guess local resources need to be packaged in the extension?

Reply:
What do you mean ?

Reply:
if a webview want to load local resources, the local resources need to be packaged in the extension.
will try to find a more concrete example.

Reply:
Yes. Let's take an exemple : 
```
src/
  extension/ --> folder with extension code, engine, ..
  webview/  --> folder with webview code
dist/
  extension/
    webview/ --> build result of the webview code
    main.js     --> entrypoint of the extension
    package.json
```
This is because when you're going to publish your extension on vscode marketplace you'll publish the extension folder, so the webview must be inside.

Reply:
ok thanks, 
If one try to load a resources (http or internal) from the code loaded in the webview ( like with a script tag), I've seen there is some limitation for external resources, but it seem one can load internal resources. But what is exactly an internal resources? is this something that need to be package within the extension?